src/test/scala/fte/data/GraphCreator.scala

```diff
@@ -0,0 +1,166 @@
+    *                 |....|....|....|
+    *                 +----+----+----+
+    *                 |....|XXXX|....|
+    *                 |....|XXXX|....|
+    *                 +----+----+----+
+    * @return The X's in the pattern definition are added to the returned graph.
+    */
+  def fromThreadDiagram(urlQuery: String): Option[Graph] = {
+    implicit val config: TilesConfig = TilesConfig(urlQuery)
+    implicit val diagram: Diagram = ThreadDiagram(NewPairDiagram.create(config))
+    implicit val scale: Int = 2
+    graphFrom(diagram.links.filter(inCenterBottomTile))
+  }
+
+  def fromPairDiagram(urlQuery: String): Option[Graph] = {
+    implicit val config: TilesConfig = TilesConfig(urlQuery)
+    implicit val diagram: Diagram = NewPairDiagram.create(config)
+    implicit val scale: Int = 1
+    graphFrom(diagram.links.filter(inCenterBottomTile))
+  }
+
+  private def graphFrom(unsortedLinks: Seq[LinkProps])
+                       (implicit diagram: Diagram, scale: Int): Option[Graph] = {
+    val linksInTile = unsortedLinks
+      .sortBy(l => diagram.node(l.target).id -> diagram.node(l.source).id)
+    val graph = new Graph()
+    val topoLinks = TopoLink.simplify(linksInTile)
+
+    // TODO for now this check prevents eternal loops for bandage/sheered pair diagrams
+    if (topoLinks.exists(l => l.sourceId == l.targetId)) return None
+
+    // create vertices
+    implicit val vertexMap: Map[String, Vertex] = topoLinks
+      .map(_.targetId).distinct.zipWithIndex
+      .map { case (nodeId, i) =>
+        // The initial coordinates don't matter as long as they are unique.
+        nodeId -> graph.createVertex(i, 0)
+      }.toMap
+    println(vertexMap.keys.toArray.sortBy(identity).mkString(","))
+
+    // create edges
+    implicit val edges: Array[Edge] = linksInTile.map { link =>
+      graph.addNewEdge(new Edge(
+        vertexMap(sourceOf(link).id),
+        vertexMap(targetOf(link).id)
+      ))
+    }.toArray
+
+    // add edges to vertices in clockwise order
+    val linksBySource = topoLinks.groupBy(_.sourceId).map { case (id, links) =>
+      id -> (links.filter(_.isRightOfSource) ++ links.filter(_.isLeftOfSource))
+    }
+    topoLinks.groupBy(_.targetId).foreach { case (id, links) =>
+      val allFour = linksBySource(id) ++ links.filter(_.isLeftOfTarget) ++ links.filter(_.isRightOfTarget)
+      //println(s"$id: " + allFour.mkString(";"))
+      allFour.map(topoLink => findEdge(topoLink).foreach(vertexMap(id).addEdge))
+    }
+
+   val edgesPerFace = facesFrom(topoLinks).map(_.counterClockWise)
+    val facesNew = edgesPerFace
+      .map(toJavaEdgeList)
+      .map(faceEdges => new Face(){setEdges(faceEdges)})
+
+    // TODO add new faces to edges (as forFace/revFace) to eliminate old faces
+    edges.foreach{ e =>
+      val fs = facesNew.filter(_.getEdges.contains(e))
+      if (fs.size != 2) print(s"whoops:${toS(e)}x${fs.size} ")
+//      e.forFace = fs.head
+//      e.revFace = fs.last
```

Comment:
Is it enough to assign each edge to two different faces, or should all edges of one face be assigned to either `forFace` or `revFace`? In other words: would the code above be enough or should I elaborate the method `Face.direction`?

The whoopses emerge on the braid/sheered patterns. At first sight the code above seems to cause an eternal loop for the first other pattern (torchon/ct) after calculation of the null space.

[log.txt](https://github.com/d-bl/GroundForge/files/4008201/log.txt) for more patterns than below:

<img width="458" alt="image" src="https://user-images.githubusercontent.com/10553630/71556060-e90f1580-2a33-11ea-819f-9ae0a158ae9e.png">

counterClockWise:
List(b20,b22,t,f, b22,a11,t,t, a11,b20,f,t, a12,b20,t,f, b21,a12,f,f, b20,b21,f,t)
List(a12,b20,t,f, b20,b21,f,t, a12,b21,t,t)
List(a11,b22,f,f, b20,b22,t,f, a11,b20,f,t)
List(b22,a10,t,f, a10,a11,f,t, b22,a11,t,t)
List(a10,a12,t,f, a12,b21,t,t, b21,a10,f,t, b22,a10,t,f, a11,b22,f,f, a10,a11,f,t)
List(b21,a12,f,f, a10,a12,t,f, b21,a10,f,t)
new:
3:4,4:2,2:3,1:3,5:1,3:5
2:4,3:4,2:3
1:3,3:5,1:5
5:1,0:1,5:0
0:1,1:5,5:0,4:0,2:4,0:2
4:0,0:2,4:2
old:
5:0,4:0,2:4,0:2,0:1,1:5
5:0,5:1,0:1
5:1,3:5,3:4,4:2,2:3,1:3
3:5,1:5,1:3
4:2,4:0,0:2
2:4,3:4,2:3

Reply:
postponing this issue a new (to be created) pull request